### PR TITLE
Add plotting features for new spin calc modes

### PR DESCRIPTION
**Pull request by ltan01**
Hello,

I added three scripts to plot the outputs for the imsigma_spin, ephmat_spin, and spins calc modes. The imsigma_spin and ephmat_spin scripts are based off of the imsigma and ephmat plotting scripts. The spins plotting script plots the <n|Sigma_z|n> values on the bands, but I tested this using regular norm and LogNorm of (1 - spin value). In this PR, the spins script uses the LogNorm.

Regular norm
si-spins

Log norm
si-spins-lognorm

### CODE REVIEW

**Comment by imaliyov** [comment]
Great work, Laurie! 
Would be great to have these plots (or similar) for the tutorials for this calc mode, perhaps, in this section: https://perturbo-code.github.io/mydoc_scattering.html

Also, are you guys planning to add this mode to PerturboPy? Ideally, all the calc modes implemented in Perturbo should be supported by PerturboPy.

Thank again!

Best,
Ivan

**Comment by imaliyov** [comment]
Great work, Laurie!

Would be also very nice to have a small tutorial on that on PerturboPy, perhaps here: https://perturbopy.readthedocs.io/en/latest/postproc/tutorials/imsigma.html

Thanks again!

Best,
Ivan

**Review by hurricane642 — CHANGES_REQUESTED** [review]
That's great work, thank you a lot!
I have comments only on the functions documentation - it should be updated according to the functions parameters and actual purpose of each function. Beside that - great work!

File: src/perturbopy/postproc/calc_modes/ephmat_spin.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+        ax = plot_vals_on_bands(ax, self.qpt.path, self.phdisp, self.phdisp.units, values=values, label=r'$\Phi$', **kwargs)
+
+        if show_qpoint_labels:
+            ax = plot_recip_pt_labels(ax, self.qpt.labels, self.qpt.points, self.qpt.path)
+
+        return ax
+
+    def plot_ephmat(self, ax, kpoint_idx=0, show_qpoint_labels=True, **kwargs):
+        """
```
[the remark this answers is not in the document]
Comment (imaliyov) [reply src/perturbopy/postproc/calc_modes/ephmat_spin.py:155]: Agreed!

File: src/perturbopy/postproc/calc_modes/spins.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+        spins_dict = self._pert_dict['spins'].pop('band index (spins)')
+        spin_units = self._pert_dict['spins'].pop('<n|Sigma_z|n> units')
+
+        self.kpt = RecipPtDB.from_lattice(kpoint, kpoint_units, self.lat, self.recip_lat, kpath, kpath_units)
+        self.bands = UnitsDict.from_dict(energies_dict, energy_units)
+        self.spins = UnitsDict.from_dict(spins_dict, spin_units)
+
+    def plot_bands(self, ax, show_kpoint_labels=True, **kwargs):
+        """
```
[the remark this answers is not in the document]
Comment (ltan01) [reply src/perturbopy/postproc/calc_modes/spins.py:58]: The energy_window parameter should still be there for both plot_bands() and plot_spins() in spins.py? I copied functions from bands.py, so the function should be identical

File: src/perturbopy/postproc/calc_modes/ephmat_spin.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        for phidx in ephmat_dat.keys():
+            phdisp[phidx] = ephmat_dat[phidx].pop('phonon energy')
+            defpot[phidx] = np.array(ephmat_dat[phidx].pop('deformation potential')).reshape(N, M)
+            ephmat[phidx] = np.array(ephmat_dat[phidx].pop('e-ph matrix elements')).reshape(N, M)
+
+        self.phdisp = UnitsDict.from_dict(phdisp, phdisp_units)
+        self.defpot = UnitsDict.from_dict(defpot, defpot_units)
+        self.ephmat = UnitsDict.from_dict(ephmat, ephmat_units)
+
+    def plot_phdisp(self, ax, show_qpoint_labels=True, **kwargs):
```
[the remark this answers is not in the document]
Comment (ltan01) [reply src/perturbopy/postproc/calc_modes/ephmat_spin.py:88]: If this function and its parameters are identical to those in the ephmat.py, how should I update the docs? Should I rewrite my scripts so I'm not repeating functions?

File: src/perturbopy/postproc/calc_modes/ephmat_spin.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ltan01) [reply src/perturbopy/postproc/calc_modes/ephmat_spin.py:149]: Added!

File: src/perturbopy/postproc/calc_modes/ephmat_spin.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        """
+        ax = plot_dispersion(ax, self.qpt.path, self.phdisp, self.phdisp.units, **kwargs)
+
+        if show_qpoint_labels:
+            ax = plot_recip_pt_labels(ax, self.qpt.labels, self.qpt.points, self.qpt.path)
+
+        return ax
+
+    def plot_defpot(self, ax, kpoint_idx=0, show_qpoint_labels=True, **kwargs):
+        """
```
[the remark this answers is not in the document]
Comment (ltan01) [reply src/perturbopy/postproc/calc_modes/ephmat_spin.py:114]: Added more here!

**Review by imaliyov — APPROVED** [review]

**Comment by imaliyov** [comment]
We have a conflict with the `src/perturbopy/postproc/calc_modes/ephmat.py` file.
Can we resolve the conflict on this branch, and then merge?

To do this, we have to `git pull origin main`, then resolve and then merge. @hurricane642 correct me if I'm wrong:)

**Comment by hurricane642** [comment]
> We have a conflict with the `src/perturbopy/postproc/calc_modes/ephmat.py` file. Can we resolve the conflict on this branch, and then merge?
> 
> To do this, we have to `git pull origin main`, then resolve and then merge. @hurricane642 correct me if I'm wrong:)

That is definitely the funniest conflict I've ever seen...
Снимок экрана 2024-12-04 в 08 21 44

Actually, @ltan01 needs to make `rebase` one more time ^^ that is the way to solve the conflicts. Then make `force push` and everything will be resolved:)

**Comment by ltan01** [comment]
> > We have a conflict with the `src/perturbopy/postproc/calc_modes/ephmat.py` file. Can we resolve the conflict on this branch, and then merge?
> > To do this, we have to `git pull origin main`, then resolve and then merge. @hurricane642 correct me if I'm wrong:)
> 
> That is definitely the funniest conflict I've ever seen... Снимок экрана 2024-12-04 в 08 21 44
> 
> Actually, @ltan01 needs to make `rebase` one more time ^^ that is the way to solve the conflicts. Then make `force push` and everything will be resolved:)

Haha, yes I’ll take care of the rebase!

**Comment by hurricane642** [comment]
Cool, seems like everything is ready. @ltan01 , feel free to merge it whenever you'll be ready!

**Pull request merged**